pkg/service/aws/oidc.go

```diff
@@ -0,0 +1,159 @@
+package aws
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/pkg/errors"
+	"gitlab.com/netbook-devs/spawner-service/pkg/config"
+	proto "gitlab.com/netbook-devs/spawner-service/proto/netbookai/spawner"
+)
+
+//arn:aws:iam::965734315247:oidc-provider
+//
+// oidc endpoint
+const trustPolicTemplate = `{
+			"Effect": "Allow",
+			"Principal": {
+				"Federated": "%s"
+			},
+			"Action": "sts:AssumeRoleWithWebIdentity",
+			"Condition": {
+				"StringEquals": {
+					"%s:sub": "system:serviceaccount:oidc:awskube2iam",
+					"%s:aud": "sts.amazonaws.com"
+				}
+			}
+		}`
+
+func getTrustPolicyDocument(federatedPrefix, oidcUrl string) string {
+	return fmt.Sprintf(trustPolicTemplate, federatedPrefix, oidcUrl, oidcUrl)
+}
+
+//generateTrustPolicyDocument read the current policy document as a map, create new policy using the template stringa and convert that to map
+//finally append the new policy to Statement section of the current policy
+func generateTrustPolicyDocument(currentPolicyDoc, oidcproviderArn, oidcIssuer string) (string, error) {
+	decodedValue, err := url.QueryUnescape(currentPolicyDoc)
+	if err != nil {
+		return "", err
+	}
+
+	policyDoc := make(map[string]interface{})
+	err = json.Unmarshal([]byte(decodedValue), &policyDoc)
+	if err != nil {
+		return "", err
+	}
+
+	newPolicy := getTrustPolicyDocument(oidcproviderArn, oidcIssuer)
+
+	newDoc := make(map[string]interface{})
+	err = json.Unmarshal([]byte(newPolicy), &newDoc)
+	if err != nil {
+		return "", err
+	}
+	policyDoc["Statement"] = append(policyDoc["Statement"].([]interface{}), newDoc)
+
+	b, err := json.Marshal(policyDoc)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// fetch OIDC from the cluster and attach it to role policy
+
+func (a *awsController) ConnectClusterOIDCToTrustPolicy(ctx context.Context, req *proto.ConnectClusterOIDCToTrustPolicyRequest) (*proto.ConnectClusterOIDCToTrustPolicyResponse, error) {
+	//get the oidc endpoint from the cluster spec
+
+	roleName := config.Get().OpenIDRole
+	region := req.Region
+	accountName := req.AccountName
+	session, err := NewSession(ctx, region, accountName)
+
+	if err != nil {
+		return nil, err
+	}
+	eksClient := session.getEksClient()
+
+	cluster, err := getClusterSpec(ctx, eksClient, req.ClusterName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get the cluster")
+	}
+
+	if *cluster.Status != eks.ClusterStatusActive {
+		a.logger.Info(ctx, "cluster is not active yet", "status", *cluster.Status)
+		return nil, fmt.Errorf("cluster is not active, curent state: %s", *cluster.Status)
+	}
+	if cluster.Identity == nil {
+		a.logger.Info(ctx, "cluster doesnt have identity", "identity", nil)
+		return nil, errors.New("cluster identity is nil")
+	}
+
+	issuer := *cluster.Identity.Oidc.Issuer
```

Comment:
Can we check if Oidc is non-nil and so is the issuer? To be on safer side?

Reply:
yeah, have added.